File: iterableapi/src/main/java/com/iterable/iterableapi/IterableLogger.java
package com.iterable.iterableapi;

import android.util.Log;

import androidx.annotation.VisibleForTesting;

import com.iterable.iterableapi.util.LogLevel;

/**
 * Created by David Truong dt@iterable.com.
 */
public final class IterableLogger {

    private IterableLogger() { }

    public static void d(String tag, String msg) {
        if (isLoggableLevel(LogLevel.DEBUG)) {
            Log.d(tag, " 💚 " + msg);
        }
    }

    public static void d(String tag, String msg, Throwable tr) {
        if (isLoggableLevel(LogLevel.DEBUG)) {
            Log.d(tag, " 💚 " + msg, tr);
        }
    }

    public static void v(String tag, String msg) {
        if (isLoggableLevel(LogLevel.VERBOSE)) {
            Log.v(tag, " 💛 " + msg);
        }
    }

    public static void w(String tag, String msg) {
        if (isLoggableLevel(LogLevel.WARN)) {
            Log.w(tag, " 🧡️ " + msg);
        }
    }

    public static void w(String tag, String msg, Throwable tr) {
        if (isLoggableLevel(LogLevel.WARN)) {
            Log.w(tag, " 🧡 " + msg, tr);
        }
    }

    public static void e(String tag, String msg) {
        if (isLoggableLevel(LogLevel.ERROR)) {
            Log.e(tag, " ❤️ " + msg);
        }
    }

    public static void e(String tag, String msg, Throwable tr) {
        if (isLoggableLevel(LogLevel.ERROR)) {
            Log.e(tag, " ❤️ " + msg, tr);
        }
    }

    public static void printInfo() {
        try {
            IterableLogger.v("Iterable Call", Thread.currentThread().getStackTrace()[3].getFileName() + " => " + Thread.currentThread().getStackTrace()[3].getClassName() + " => " + Thread.currentThread().getStackTrace()[3].getMethodName() + " => Line #" + Thread.currentThread().getStackTrace()[3].getLineNumber());
        } catch (Exception e) {
            IterableLogger.e("Iterable Call", "Couldn't print info");
        }
    }

    @VisibleForTesting
    protected static boolean isLoggableLevel(@LogLevel.Level int messageLevel) {
        return messageLevel >= getLogLevel();
    }

    @LogLevel.Level
    private static int getLogLevel() {
        if (IterableApi.sharedInstance != null) {
            if (IterableApi.sharedInstance.getDebugMode()) {
                return LogLevel.VERBOSE;
            } else {
                return IterableApi.sharedInstance.config.logLevel;
            }
        }
        return LogLevel.NONE;
    }
}

File: iterableapi/src/main/java/com/iterable/iterableapi/IterableLogger.kt
package com.iterable.iterableapi

import android.util.Log

/**
 * Created by David Truong dt@iterable.com.
 */
object IterableLogger {
    private const val TAG = "IterableApi"

    fun v(tag: String, msg: String) {
        if (IterableApi.getInstance().getDebugMode()) {
            Log.v(tag, msg)
        }
    }

    fun d(tag: String, msg: String) {
        if (IterableApi.getInstance().getDebugMode()) {
            Log.d(tag, msg)
        }
    }

    fun i(tag: String, msg: String) {
        if (IterableApi.getInstance().getDebugMode()) {
            Log.i(tag, msg)
        }
    }

    fun w(tag: String, msg: String) {
        if (IterableApi.getInstance().getDebugMode()) {
            Log.w(tag, msg)
        }
    }

    fun w(tag: String, msg: String, tr: Throwable) {
        if (IterableApi.getInstance().getDebugMode()) {
            Log.w(tag, msg, tr)
        }
    }

    fun e(tag: String, msg: String) {
        if (IterableApi.getInstance().getDebugMode()) {
            Log.e(tag, msg)
        }
    }

    fun e(tag: String, msg: String, tr: Throwable) {
        if (IterableApi.getInstance().getDebugMode()) {
            Log.e(tag, msg, tr)
        }
    }

    fun printInfo() {
        if (IterableApi.getInstance().getDebugMode()) {
            val stackTrace = Thread.currentThread().stackTrace
            if (stackTrace.size >= 4) {
                val callingClass = stackTrace[3].className.split(".").last()
                val callingMethod = stackTrace[3].methodName
                d(callingClass, callingMethod)
            }
        }
    }
}
